[libc++][type_traits] Implements "A type trait to detect reference binding to temporary"

H-G-Hristov opened the pull request:
Implements partially: [P2255R2: A type trait to detect reference binding to temporary](https://wg21.link/P2255R2)
Issue: https://github.com/llvm/llvm-project/issues/105180

https://eel.is/c++draft/meta.type.synop
https://eel.is/c++draft/meta.unary.prop

Implented type traits:
- [x] `reference_constructs_from_temporary`
- [x] `reference_converts_from_temporary`

Closes #129049

Minor drive-by tweak to `std::is_implicit_lifetime` tests.

H-G-Hristov committed:
[libc++][type_traits] Implements "A type trait to detect reference binding to temporary"

Implements partially: P2255R2: A type trait to detect reference binding to temporary
https://github.com/llvm/llvm-project/issues/105180

https://eel.is/c++draft/meta.type.synop
https://eel.is/c++draft/meta.unary.prop

Implented type traits:
- [x] `reference_constructs_from_temporary`
- [x] `reference_converts_from_temporary`

frederick-vs-ja commented on libcxx/include/__type_traits/reference_constructs_from_temporary.h:
```diff
@@ -0,0 +1,42 @@
+#if !defined(_LIBCPP_CXX03_LANG)
+
+template <class _Tp, class _Up>
+inline constexpr bool __reference_constructs_from_temporary_v = __reference_constructs_from_temporary(_Tp, _Up);
```
I think `const` can be used instead to make this usable in C++03 mode. `__reference_constructs_from_temporary_v<T, U>` will still be usable in constant expressions as it's of type `const bool`.

H-G-Hristov replied:
I might have been confused here. Do you mean the frozen C++03 headers have nothing to do with the C++03 mode?

frederick-vs-ja replied:
I'm not sure. Perhaps the "C++03 with non-frozen headers" mode is still being tested.

H-G-Hristov replied:
You are right. This makes sense. Thanks.

frederick-vs-ja commented on libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/reference_constructs_from_temporary.pass.cpp:
```diff
@@ -0,0 +1,121 @@
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: std-at-least-c++23
+
```
I _think_ we should require Clang 18 here. GCC is still unsupported.

frederick-vs-ja replied:
> If this is supposed to land before LLVM20 is released we should support Clang 18 but after that we don't. Right?

Hmm... I'm not sure whether some versions of Apple Clang should be considered unsupported.

> What do you mean "GCC is still unsupported"? It looks like GCC14 implemented [P2255R2](https://wg21.link/P2255R2), although I see a compilation error in our tests.

Sorry for misleading... It seems that GCC 14 has supported these 2 intrinsics, but access control was incorrectly ignored. So perhaps we should just skip the `ConvertsToRefPrivate` cases for GCC now.

frederick-vs-ja commented on libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/reference_converts_from_temporary.pass.cpp:
```diff
@@ -0,0 +1,124 @@
+// REQUIRES: std-at-least-c++23
+
+// These compilers don't support std::reference_converts_from_temporary yet.
+// UNSUPPORTED: android, apple-clang-15, apple-clang-16.1, apple-clang-16.2, clang-19
```
The `__reference_converts_from_temporary` intrinsic landed in clang-19. Why isn't clang-19 supported?

H-G-Hristov replied:
It looks like `__has_builtin(__reference_converts_from_temporary)` returns `false` in Clang 19, although I think the intrinsic is available. 
https://godbolt.org/z/rjoa48caa

frederick-vs-ja replied:
> It looks like `__has_builtin(__reference_converts_from_temporary)` returns `false` in Clang 19, although I think the intrinsic is available.
> https://godbolt.org/z/rjoa48caa

Ah, this is #111477. Perhaps workaround should be used.

H-G-Hristov replied:
What would be an appropriate workaround?

frederick-vs-ja replied:
Maybe we should use `__has_builtin(__reference_converts_from_temporary) || (defined(_LIBCPP_COMPILER_CLANG_BASED) && __clang__ >= 19)` (or `18` for `__reference_constructs_from_temporary`). I'm not sure how to handle Apple Clang, though.

H-G-Hristov replied:
AFAIK. `__has_builtin(__reference_converts_from_temporary)` should work in Clang 18, it is broken on Clang 19.1.1 (what I use right now) it should be fixed with this patch
https://github.com/llvm/llvm-project/commit/dedbdfb70daf06fc07bb4dfbec1d18af8fcef28f
What is not clear to me, was it fixed in Clang 19.1.2 or in Clang 19.1.7.

```C++
#if __has_builtin(__reference_converts_from_temporary) || (defined(_LIBCPP_COMPILER_CLANG_BASED) && (__clang_major == 18 || (__clang_major >= 19 && __clang_minor >= 2))
...
```
I don't have access to macOS any longer. Xcode 16.3 should be based on LLVM19 but I guess we won't get the fix before Xcode 17.

H-G-Hristov committed:
Address comments

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Try to fix CI

H-G-Hristov committed:
Fixed modules map condition

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Fixed formatting

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Try to fix CI

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Try to fix Apple CI

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Try to fix CI

H-G-Hristov committed:
Fixed formatting

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Try to fix Android

H-G-Hristov committed:
Merge branch 'hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary' of https://github.com/H-G-Hristov/llvm-project into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Merge branch 'llvm:main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

frederick-vs-ja commented on libcxx/include/__type_traits/reference_converts_from_temporary.h:
```diff
@@ -0,0 +1,44 @@
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if __has_builtin(__reference_converts_from_temporary) ||                                                              \
+    (defined(_LIBCPP_COMPILER_CLANG_BASED) && (__clang_major__ >= 19) && (__clang_minor__ >= 1) &&                     \
+     (__clang_patchlevel__ >= 2))
```
I think it would be clearer to move the conditions to `<__config>` and create a temporary macro.
Also, couldn't we enable the code for Clang 19.1.0? IIUC in that version `__has_builtin(__reference_converts_from_temporary)` was wrong but `__reference_converts_from_temporary` already worked.

H-G-Hristov replied:
Nice idea, thank you!

H-G-Hristov replied:
Android Clang vs LLVM relation can be checked here (Base revision):

https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r547379/
- [clang_source_info.md](https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r547379/clang_source_info.md)

https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r536225/
- [clang_source_info.md](https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r536225/clang_source_info.md)

https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r530567/
- [clang_source_info.md](https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r530567/clang_source_info.md)

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Addressed review comments

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Try to fix CI

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Merge branch 'hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary' of https://github.com/H-G-Hristov/llvm-project into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Minor tweaks

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Fixed formatting

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Enable ANDROID

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Updated tests

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Fix C++03

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

Zingam committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

frederick-vs-ja commented on libcxx/include/__type_traits/reference_constructs_from_temporary.h:
```diff
@@ -0,0 +1,42 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___TYPE_TRAITS_REFERENCE_CONSTRUCTS_FROM_TEMPORARARY_H
```
The macro name (and `_LIBCPP___TYPE_TRAITS_REFERENCE_CONVERTS_FROM_TEMPORARARY_H`) looks a bit weird.
```suggestion
#ifndef _LIBCPP___TYPE_TRAITS_REFERENCE_CONSTRUCTS_FROM_TEMPORARY_H
```

frederick-vs-ja commented on libcxx/include/__config:
```diff
@@ -1215,6 +1215,21 @@ typedef __char32_t char32_t;
 
+// TODO(LLVM 22): Remove this macro once LLVM19 support ends.
+#  if __has_builtin(__reference_constructs_from_temporary) ||                                                          \
+      (defined(_LIBCPP_CLANG_VER) && (_LIBCPP_CLANG_VER >= 1900))
```
No change requested. I still think that this should be `_LIBCPP_CLANG_VER >= 1800`. But as support for Clang 18 will be dropped in LLVM 21 release, so the difference only affects CI, IIUC.

Zingam replied:
`#if __has_builtin(__reference_constructs_from_temporary)" works correctly in Clang 18, so I think the above expresses the workaround more clearly.
https://godbolt.org/z/f8a1MG1Gq
If you think it's better to do it as you suggest. I'll do it. No problem.

I think the only case that's not handled completely currently is `__reference_constructs_from_temporary` on Android since we don't have a compiler that implements it on that platform and I suspect that the next NDK won't include the support too.

frederick-vs-ja replied:
Yeah. The current conditions look fine to me with your explanation.

frederick-vs-ja reviewed (commented):
I think we're very near to landing this. Release notes should be added.

H-G-Hristov committed:
Fixed type in macro name

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Android NDK r29 support + Release Notes

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov commented:
> I think we're very near to landing this. Release notes should be added.

I did a few  changes:
- Added a note to Release Notes.
- Updated the code and tests for Android NDK r29, which was just released: https://github.com/android/ndk/wiki/Changelog-r29. The compiler is: https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r547379/ and reports as Clang 20.0, which features `__reference_converts_from_temporary` but not the fix for `__has_builtin(__reference_converts_from_temporary)`. With the latest change this upcoming version should be handled now too.

frederick-vs-ja commented on libcxx/include/__config:
```diff
@@ -1215,6 +1215,21 @@ typedef __char32_t char32_t;
+#  endif
+
+// TODO(LLVM 22): Remove this macro once LLVM19 support ends.
+#  if __has_builtin(__reference_converts_from_temporary) || (defined(_LIBCPP_CLANG_VER) && (_LIBCPP_CLANG_VER >= 1901))
```
I wonder why is `1901` but not `1900` used.

frederick-vs-ja commented on libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/reference_converts_from_temporary.pass.cpp:
```diff
@@ -0,0 +1,90 @@
+// REQUIRES: std-at-least-c++23
+
+// These compilers don't support std::reference_converts_from_temporary yet.
+// UNSUPPORTED: android && clang-19.0, apple-clang-15, apple-clang-16, clang-18
```
Is the `android && clang-19.0` configuration not using a "true" Clang 19? I think this configuration should be supported.

Zingam replied:
Please check the comment above: https://github.com/llvm/llvm-project/pull/128649#discussion_r1982780610

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Updated macros to handle ANDROID clearer

frederick-vs-ja reviewed (approved):
I think we can land this now.

philnik777 commented on libcxx/include/__config:
```diff
@@ -1215,6 +1215,24 @@ typedef __char32_t char32_t;
 #    define _LIBCPP_HAS_EXPLICIT_THIS_PARAMETER 0
 #  endif
 
+// TODO(LLVM 22): Remove this macro once LLVM19 support ends.
+#  if __has_builtin(__reference_constructs_from_temporary) ||                                                          \
+      (defined(_LIBCPP_CLANG_VER) &&                                                                                   \
+       ((!defined(__ANDROID__) && _LIBCPP_CLANG_VER >= 1901) || (defined(__ANDROID__) && _LIBCPP_CLANG_VER >= 1900)))
+#    define _LIBCPP_HAS_REFERENCE_CONSTRUCTS_FROM_TEMPORARY 1
+#  else
+#    define _LIBCPP_HAS_REFERENCE_CONSTRUCTS_FROM_TEMPORARY 0
+#  endif
```
This should just be `__has_builtin(__reference_constructs_from_temporary)`. 99% of people use the same version of Clang and libc++, and this is way too complicated of a condition compared to what we gain.

philnik777 commented on libcxx/include/__type_traits/reference_constructs_from_temporary.h:
```diff
@@ -0,0 +1,42 @@
+
+#if _LIBCPP_HAS_REFERENCE_CONSTRUCTS_FROM_TEMPORARY
+
+template <class _Tp, class _Up>
+const bool __reference_constructs_from_temporary_v = __reference_constructs_from_temporary(_Tp, _Up);
```
We don't need to introduce private names if we don't actually use them.

frederick-vs-ja replied:
Hmm, it seems better to drop the private names for now. We can still add them when implementing [LWG4148](https://cplusplus.github.io/LWG/issue4148) (in old modes). CC @elhewaty.

H-G-Hristov committed:
Addressed review comments

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov commented:
Removed workarounds and disabled tests for unsupported compilers: 
> According to https://github.com/llvm/llvm-project/issues/111477 `__has_builtin(`reference_{constructs|converts|_from_temporary) was broken in Clang 19.1.0 and it was fixed in [Clang 19.1.2](https://github.com/llvm/llvm-project/pull/111660#issuecomment-2479496809) with https://github.com/llvm/llvm-project/commit/dedbdfb70daf06fc07bb4dfbec1d18af8fcef28f. Workarounds were implemented and test for unsupported compilers disabled.

H-G-Hristov committed:
Addressed review comments

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov commented:
> If only the check was broken, can't we just drop the check entirely?

Wouldn't this cause build failures in the CI if the the intrinsic isn't available?

philnik777 commented:
> > If only the check was broken, can't we just drop the check entirely?
> 
> Wouldn't this cause build failures in the CI if the the intrinsic isn't available?

When is it not available?

H-G-Hristov commented:
> > > If only the check was broken, can't we just drop the check entirely?
> > 
> > 
> > Wouldn't this cause build failures in the CI if the the intrinsic isn't available?
> 
> When is it not available?

On Arm currently, Apple Clang

H-G-Hristov committed:
Fix CI

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

frederick-vs-ja commented on libcxx/include/type_traits:
```diff
@@ -524,6 +533,9 @@ namespace std
 #  if _LIBCPP_STD_VER >= 23
 #    include <__type_traits/is_implicit_lifetime.h>
 #  endif
+// C++23 but private declarations are available in previous standards.
+#  include <__type_traits/reference_constructs_from_temporary.h>
+#  include <__type_traits/reference_converts_from_temporary.h>
```
I think at this moment we should guard this with `#  if _LIBCPP_STD_VER >= 23`. Let's defer the exposure of private declarations to the PR for LWG4148.

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Review comments

frederick-vs-ja commented on libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/common.h:
```diff
@@ -39,9 +39,51 @@ struct A {
+
+#if TEST_STD_VER >= 23
+
+struct NonPOD {
+  NonPOD(int);
+};
+enum Enum { EV };
+struct POD {
+  Enum e;
+  int i;
+  float f;
+  NonPOD* p;
+};
+// Not PODs
+struct Derives : POD {};
```
These type names are confusing. Perhaps we only want to use `Base` and `Derived`.

frederick-vs-ja reviewed (approved):
LGTM!

Some stuffs in the test code are confusing, but I think we can do the cleanup later.

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2255R2-A_type_trait_to_detect_reference_binding_to_temporary

H-G-Hristov committed:
Renamed types

frederick-vs-ja commented:
I guess we need to revert the drive-by changes. Not sure whether that would be temporary.

At the time of last CI for this PR, the "macos (apple-configuration, macos-15)" configuration was using apple-clang-16. But today it changed to use apple-clang-17 and the tests for `is_implicit_lifetime` still failed.

philnik777 commented on libcxx/include/__type_traits/reference_constructs_from_temporary.h:
```diff
@@ -0,0 +1,35 @@
+    : public bool_constant<__reference_constructs_from_temporary(_Tp, _Up)> {};
+
+template <class _Tp, class _Up>
+_LIBCPP_NO_SPECIALIZATIONS inline constexpr bool reference_constructs_from_temporary_v =
```
@Zingam It looks like you forgot to add tests for the `_LIBCPP_NO_SPECIALIZATIONS` annotations. Could you add them in a follow-up PR?

Zingam replied:
Thanks! I'll do it.
